### PR TITLE
Use `sysadminctl` instead of `dscl`

### PR DESCRIPTION
**Pull request by Enzime**
Extracted from #1017

This should be PR should be a no-op

### CODE REVIEW

File: modules/users/default.nix [hunk 1]
```diff
@@ -182,7 +184,7 @@ in
         if [ -n "$u" ]; then
           if [ "$u" -gt 501 ]; then
             echo "deleting user ${name}..." >&2
-            dscl . -delete '/Users/${name}' 2> /dev/null
+            sysadminctl -deleteUser '${name}' 2> /dev/null
```
Comment (Enzime) [inline modules/users/default.nix:187]:
```suggestion
            sysadminctl -deleteUser '${name}'
```

Need to check if the `2> /dev/null` is still necessary

Comment (Enzime) [reply modules/users/default.nix:187]: I'll keep it as `sysadminctl` outputs a lot of unnecessary stuff to stderr

**Comment by Enzime** [comment]
@emilazy this is ready to review, the current test failure is unrelated

**Review by emilazy — CHANGES_REQUESTED** [review]
Broadly LGTM, thanks! I haven’t tested adding and deleting users with this yet, so one of us probably should before merging if you haven’t already. Some nits.

File: modules/users/default.nix [hunk 2]
```diff
@@ -147,28 +147,28 @@ in
 
       ${concatMapStringsSep "\n" (v: ''
         ${optionalString cfg.forceRecreate ''
-          u=$(dscl . -read '/Users/${v.name}' UniqueID 2> /dev/null) || true
-          u=''${u#UniqueID: }
+          u=$(id -u ${lib.escapeShellArg v.name} 2> /dev/null) || true
           if [[ "$u" -eq ${toString v.uid} ]]; then
             echo "deleting user ${v.name}..." >&2
-            dscl . -delete '/Users/${v.name}' 2> /dev/null
+            sysadminctl -deleteUser '${v.name}' 2>/dev/null
```
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/default.nix]: Done 👍

File: modules/users/default.nix [hunk 3]
```diff
@@ -147,28 +147,28 @@ in
...
           if [ -z "$u" ]; then
             echo "creating user ${v.name}..." >&2
-            dscl . -create '/Users/${v.name}' UniqueID ${toString v.uid}
-            dscl . -create '/Users/${v.name}' PrimaryGroupID ${toString v.gid}
+            sysadminctl -addUser '${v.name}' \
+              -UID ${toString v.uid} \
+              -GID ${toString v.gid} \
+              -fullName '${v.description}' \
+              -home '${v.home}' \
+              -shell ${lib.escapeShellArg (shellPath v.shell)}
```
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/default.nix]: Done 👍

File: modules/users/default.nix [hunk 4]
```diff
@@ -177,12 +177,11 @@ in
...
 
       ${concatMapStringsSep "\n" (name: ''
-        u=$(dscl . -read '/Users/${name}' UniqueID 2> /dev/null) || true
-        u=''${u#UniqueID: }
+        u=$(id -u ${lib.escapeShellArg name} 2> /dev/null) || true
         if [ -n "$u" ]; then
           if [ "$u" -gt 501 ]; then
             echo "deleting user ${name}..." >&2
-            dscl . -delete '/Users/${name}' 2> /dev/null
+            sysadminctl -deleteUser '${name}' 2> /dev/null
```
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/default.nix]: Done 👍

File: modules/users/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/default.nix]: Done 👍

File: modules/users/group.nix [hunk 5]
```diff
@@ -1,11 +1,10 @@
 { name, lib, ... }:
 
-with lib;
-
 {
-  options = {
+  options = with lib; {
```
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/group.nix]: Done 👍

File: tests/users-groups.nix [hunk 6]
```diff
@@ -22,42 +22,45 @@
...
-    grep "dscl . -create '/Users/foo' PrimaryGroupID 42000" ${config.out}/activate
-    grep "dscl . -create '/Users/foo' IsHidden 0" ${config.out}/activate
-    grep "dscl . -create '/Users/foo' RealName 'Foo user'" ${config.out}/activate
-    grep "dscl . -create '/Users/foo' NFSHomeDirectory '/Users/foo'" ${config.out}/activate
-    grep "dscl . -create '/Users/foo' UserShell ${lib.escapeShellArg "/run/current-system/sw/bin/bash"}" ${config.out}/activate
-    grep "dscl . -create '/Users/created.user' UniqueID 42001" ${config.out}/activate
-    grep "dscl . -create '/Users/created.user' UserShell ${lib.escapeShellArg "/sbin/nologin"}" ${config.out}/activate
+    # checking user creation in /activate
+    # REMOVEME when 24.05 is EOL: `lib.escapeShellArg` conditionally adds quotes since 24.11
+    grep -zoP "sysadminctl -addUser 'foo' (.|\n)* -UID 42000 (.|\n)* -GID 42000 (.|\n)* -fullName 'Foo user' (.|\n)* -home '/Users/foo' (.|\n)* -shell ${lib.escapeShellArg "/run/current-system/sw/bin/bash"}" ${config.out}/activate
```
[the remark this answers is not in the document]
Comment (Enzime) [reply tests/users-groups.nix]: It was previously skipping the backslash and newline however I've replaced the run-on command with `lib.escapeShellArgs` so `(.|\n)*` is no longer necessary

File: tests/users-groups.nix [hunk 7]
```diff
@@ -22,42 +22,45 @@
   users.users."unknown.user".uid = 42002;
 
   test = ''
-    echo "checking group creation in /activate" >&2
+    set -v
```
[the remark this answers is not in the document]
Comment (Enzime) [reply tests/users-groups.nix:25]: I left this in as the test otherwise doesn't output very much and is quite cryptic on failure

**Comment by Enzime** [comment]
@emilazy I've tested this now and it looks like it's working, it should be a relatively safe change as it's still all behind `users.knownUsers` and `users.knownGroups`

**Review by emilazy — APPROVED** [review]
LGTM, will let you decide whether to merge as‐is or apply @isabelroses’ suggestions.

File: modules/users/default.nix [hunk 8]
```diff
@@ -96,45 +95,49 @@ in
     system.activationScripts.groups.text = mkIf (cfg.knownGroups != []) ''
       echo "setting up groups..." >&2
 
-      ${concatMapStringsSep "\n" (v: ''
+      ${concatMapStringsSep "\n" (v: let
```
[the remark this answers is not in the document]
Comment (Enzime) [reply modules/users/default.nix:98]: I appreciate the suggestion but in the end it'll require calling `map` as well, from what I can tell the main benefit of `concatLines` is that it ensures a trailing newline which doesn't make a lot of difference here, so I'm choosing to leave it as is

**Pull request merged**